syntax_and_semantics/inheritance.md

```diff
@@ -102,3 +102,57 @@ end
 ```
 
 Without arguments or parentheses, `super` receives the same arguments as the method's arguments. Otherwise, it receives the arguments you pass to it.
+
+## Covariance and Contravariance
+
+One place inheritance can get a little tricky is with arrays. We have to be careful when declaring an array of objects where inheritance is used. For example, consider the following
+
+```crystal
+class Foo
+end
+
+class Bar < Foo
+end
+
+foo_arr = [Bar.new] of Foo # => [#<Bar:0x10215bfe0>] : Array(Foo)
+bar_arr = [Bar.new] # => [#<Bar:0x10215bfd0>] : Array(Bar)
+bar_arr2 = [Foo.new] of Bar # compiler error
+```
+
+A Foo array can hold both Foo's and Bar's, but an array of Bar can only hold Bar and it's subclasses.
+
+One place this might trip you up is when automatic casting comes into play. For example, the following won't work:
+
+```crystal
+class Foo
+end
+
+class Bar < Foo
+end
+
+class Test
+  @arr : Array(Foo)
+  def initialize
+    @arr = [Bar.new]
+  end
+end
+```
+
+because in the initialize the default type for @arr is `Array(Bar)` but the required type is `Array(Foo)`. You can solve this by specifying the type explicitly:
```

Comment:
> default type for @arr

This is not concept of the type system. I would suggest explaining that the type of the `[Bar.new]` expression is `Array(Bar)` and that is not assignable to an `Array(Foo)` instance var.

Then it follows from the previous example why the expression `[Bar.new] of Foo` works.